### PR TITLE
Fees apply dev

### PR DESCRIPTION
**Pull request by RostyslavBortmanAdoria**
All possible optimizations made. Internal check

### CODE REVIEW

File: contracts/TokenIOLib.sol [hunk 1]
```diff
@@ -878,8 +869,7 @@ library TokenIOLib {
    * @return { "issuerFirm" : "Name of the firm registered to authority" }
    */
   function getFirmFromAuthority(Data storage self, address authorityAddress) internal view returns (string issuerFirm) {
-    bytes32 id = keccak256(abi.encodePacked('registered.authority.firm', getForwardedAccount(self, authorityAddress)));
-    return self.Storage.getString(id);
+    return self.Storage.getFirmFromAuthority(authorityAddress);
```
Comment (vlad-kostanda) [inline contracts/TokenIOLib.sol]: getForwardedAccount seems missing?

File: contracts/TokenIOLib.sol [hunk 2]
```diff
@@ -1222,5 +1209,24 @@ library TokenIOLib {
...
+        bytes memory b = bytes(_b);
+        uint minLength = a.length;
+        if (b.length < minLength) minLength = b.length;
+        //@todo unroll the loop into increments of 32 and do full 32 byte comparisons
+        for (uint i = 0; i < minLength; i ++)
+            if (a[i] < b[i])
+                return false;
+            else if (a[i] > b[i])
+                return false;
+        if (a.length < b.length)
```
Comment (vlad-kostanda) [inline contracts/TokenIOLib.sol]: length should be checked before the loop and can be done with one comparison, no?
the comparison doesn't look effective in general

File: contracts/TokenIOStorage.sol [hunk 3]
```diff
@@ -637,7 +646,7 @@ contract TokenIOStorage is Ownable {
      * @param _fxUSDBPSRate  usdbps rate
      * @return { "uint" : "Returns the uint value associated with the key" }
      */
-    function setTokenfxUSDBPSRate(address _address, uint _fxUSDBPSRate) external view returns(bool success) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol:640]: how it could modify the state before, with "view"?

File: contracts/TokenIOStorage.sol [hunk 4]
```diff
@@ -647,9 +656,87 @@ contract TokenIOStorage is Ownable {
      * @param _address Pointer identifier for value in mapping
      * @return { "uint" : "Returns the uint value associated with the key" }
      */
-    function deleteTokenfxUSDBPSRate(address _address) external view returns(bool success) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol:650]: was missing onlyOwner a bug? Need to doublecheck all similar methods for permissions

File: contracts/TokenIOStorage.sol [hunk 5]
```diff
@@ -647,9 +656,87 @@ contract TokenIOStorage is Ownable {
...
     }
 
+
+    function setFirm(string firmName, bool approved) external onlyOwner returns (bool success) {
+      firms.push(Firm(firmName, address(0), approved));
+      return true;
+    }
+
+    function getFirmDetails(string firmName) external view returns(address authorityAddress, bool approved) {
+        for(uint i = 0; i < firms.length; i++) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol]: undefined amount of gas used. Potential point of failure. Let's avoid this array entirely

File: contracts/TokenIOStorage.sol [hunk 6]
```diff
@@ -647,9 +656,87 @@ contract TokenIOStorage is Ownable {
...
+            if(compare(firms[i].firmName, firmName)) {
+                authorityAddress = firms[i].authority;
+                approved = firms[i].authorized;
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol]: loop not left after the firm is found?!

File: contracts/TokenIOStorage.sol [hunk 7]
```diff
@@ -647,9 +656,87 @@ contract TokenIOStorage is Ownable {
...
+     * @param firmName Name of firm
+     * @param authority Address of authority account
+     * @param _authorized Authorization status
+     * @return {"success" : "Returns true if lib.setRegisteredAuthority succeeds"}
+     */
+    function setRegisteredAuthority(string firmName, address authority, bool _authorized) external onlyOwner returns (bool success) {
+        for(uint i = 0; i < firms.length; i++) {
+            if(compare(firms[i].firmName, firmName)) {
+                firms[i].authority = authority;
+                firms[i].authorized = _authorized;
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol]: same not returning here

File: contracts/TokenIOStorage.sol [hunk 8]
```diff
@@ -647,9 +656,87 @@ contract TokenIOStorage is Ownable {
...
+            }
+        }
+        return true;
+    }
+
+    function compare(string _a, string _b) internal pure returns (bool) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol]: you can compare strings just using keccak256, right?

File: contracts/TokenIOLib.sol [hunk 9]
```diff
@@ -148,8 +148,8 @@ library TokenIOLib {
    * @param flatFee Flat fee for interface contract transactions
    * @return {"success" : "Returns true when successfully called from another contract"}
    */
-  function setFees(Data storage self, uint maxFee, uint minFee, uint bpsFee, uint flatFee) internal returns (bool success) {
-      return self.Storage.setFees(address(this), maxFee, minFee, bpsFee, flatFee);
+  function setFees(Data storage self, address issuer, uint maxFee, uint minFee, uint bpsFee, uint flatFee) internal returns (bool success) {
+      return self.Storage.setFees(issuer, maxFee, minFee, bpsFee, flatFee);
```
Comment (vlad-kostanda) [inline contracts/TokenIOLib.sol]: is it a preparation for proxy?

File: contracts/TokenIOStableSwap.sol [hunk 10]
```diff
@@ -181,17 +181,7 @@ contract TokenIOStableSwap is Ownable {
...
-
-    require(lib.Storage.setUint(keccak256(abi.encodePacked('fee.min', asset)), feeMin),
-      'Error: Failed to set fee parameters with storage contract. Please check permissions.');
-
-    require(lib.Storage.setUint(keccak256(abi.encodePacked('fee.bps', asset)), feeBps),
-      'Error: Failed to set fee parameters with storage contract. Please check permissions.');
-
-    require(lib.Storage.setUint(keccak256(abi.encodePacked('fee.flat', asset)), feeFlat),
-      'Error: Failed to set fee parameters with storage contract. Please check permissions.');
+    require(lib.setFees(address(this), feeMax, feeMin, feeBps, feeFlat), "Error: Unable to set fee contract settings");
```
Comment (vlad-kostanda) [inline contracts/TokenIOStableSwap.sol]: asset, not address(this), no?

File: contracts/TokenIOStorage.sol [hunk 11]
```diff
@@ -646,7 +646,7 @@ contract TokenIOStorage is Ownable {
      * @param _fxUSDBPSRate  usdbps rate
      * @return { "uint" : "Returns the uint value associated with the key" }
      */
-    function setTokenfxUSDBPSRate(address _address, uint _fxUSDBPSRate) external returns(bool success) {
+    function setTokenfxUSDBPSRate(address _address, uint _fxUSDBPSRate) external onlyOwner returns(bool success) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOStorage.sol:640]: ok, who is the owner of the storage contract? And where other such permission modifiers could be missing?

File: contracts/TokenIOFeeContract.sol [hunk 12]
```diff
@@ -62,7 +62,7 @@ contract TokenIOFeeContract is Ownable {
 	 * returns {"success" : "Returns true if successfully called from another contract"}
 	 */
 	function setFeeParams(uint feeBps, uint feeMin, uint feeMax, uint feeFlat, bytes feeMsg) public onlyOwner returns (bool success) {
-		require(lib.setFees(feeMax, feeMin, feeBps, feeFlat), "Error: Unable to set fee contract settings");
+		require(lib.setFees(address(this), feeMax, feeMin, feeBps, feeFlat), "Error: Unable to set fee contract settings");
```
Comment (vlad-kostanda) [inline contracts/TokenIOFeeContract.sol:65]: is it a bug fix?

**Review by vlad-kostanda — CHANGES_REQUESTED** [review]
a no go for now, let's fix and improve. string comparison and arrays manipulations look sloppy, consume gas depending on the array size, etc.

File: contracts/TokenIOLib.sol (hunk 1 above)
Comment (RostyslavBortman) [reply contracts/TokenIOLib.sol]: Fixed

File: contracts/TokenIOLib.sol (hunk 2 above)
Comment (RostyslavBortman) [reply contracts/TokenIOLib.sol]: Fixed

File: contracts/TokenIOStorage.sol (hunk 4 above)
Comment (RostyslavBortman) [reply contracts/TokenIOStorage.sol:650]: yes, it was a bug 
I already found I think all bugs. It related only to this fxSDBPSRat set/delete.

File: contracts/TokenIOStorage.sol (hunk 5 above)
Comment (RostyslavBortman) [reply contracts/TokenIOStorage.sol]: Yes it's a concept problem. Need to discuss

File: contracts/TokenIOLib.sol [hunk 13]
```diff
@@ -148,8 +148,8 @@ library TokenIOLib {
    * @param flatFee Flat fee for interface contract transactions
    * @return {"success" : "Returns true when successfully called from another contract"}
    */
-  function setFees(Data storage self, uint maxFee, uint minFee, uint bpsFee, uint flatFee) internal returns (bool success) {
-      return self.Storage.setFees(address(this), maxFee, minFee, bpsFee, flatFee);
+  function setFees(Data storage self, address _key, uint maxFee, uint minFee, uint bpsFee, uint flatFee) internal returns (bool success) {
```
Comment (vlad-kostanda) [inline contracts/TokenIOLib.sol]: let's avoid "key" as a special word, sounds like something crypto related. change to maybe "interface"?

**Review by vlad-kostanda — CHANGES_REQUESTED** [review]
just change the naming

**Pull request closed without merging**